### PR TITLE
[PR] Adding `WYSIWYG` initial implementation

### LINKED ISSUE #1: Feat: Write a comprehensive `WYSIWYG` Tutorial for `Flutter` 📝
This is going to be one of our biggest standalone tutorials yet but it needs to be _comprehensive_ because we are going to use what we learn from this _immediately_ in our `App`. ref: https://github.com/dwyl/app/issues/275

Our `WYSIWYG` requirements for the `App` are actually quite basic https://github.com/dwyl/app/issues/275#issuecomment-1646862277 
but we need to be able to add our own **_custom_ buttons+actions** so please have that in mind when researching. 🔍  

# Todo

+ [x] Create a new `Flutter` App. 🆕 
+ [x] Research and share the available `WISYWIG` libraries. 🔍 
+ [x] Implement the most promising lib that will allow us to meet our objectives:
  + [x] **Basics**: **`B`**, _`I`_ and `U` (underline) 📝 
  + [x] Upload an in-line an image; ref: https://github.com/dwyl/flutter-image-upload-demo 🖼️ 
  + [ ] ***Advanced***: Create our own **_Custom_ Actions/Buttons** 

The _single_ most important requirement that won't immediately be clear from the the `WYSIWIG` title 
is actually _how_ the `data` gets ***transformed*** when you apply an action. 🪄 
See: https://github.com/dwyl/deltamin

The **`USP`** of what we are building is to have the _full_ history of every action that is taken
thus having full "**undo**" capability and  "***replay-ability***". 
Therefore we need to not _just_ have a `WISYWIG` lib, we need one that gives us [Operational Transform](https://en.wikipedia.org/wiki/Operational_transformation)
**So _that_** we can _easily_ store the "story" of what has happened to an `item`
and can _replay_ it both on the _connected_ devices and to _future_ team members
when they need an answer to the question: "**_How did we get `here`_**?"  

Which is why I'm leaning toward: https://github.com/singerdmx/flutter-quill or https://pub.dev/packages/quill_html_editor
which both implement [`Quill`](https://quilljs.com/docs/formats) which has [`Delta`](https://quilljs.com/docs/delta/).

Yes, I _know_ that [`YJS`](https://github.com/nelsonic/amemo/issues/62) is the `new` "hotness" in this space.
The demos are sick: https://yjs.dev/#demos
But there isn't anything in `Dart` we can use for our `Flutter App` right now ... 
Given that `Delta` is just a description language, it should be fairly easy to port to anything `else` in the future ... 💭 
So I don't have any "concerns" with using it for now.

### PR DESCRIPTION
**Pull request by LuchoTurtle**
closes #1 

This adds an implementation example using https://github.com/singerdmx/flutter-quill and documents the options available in order to fulfill the requirements stated in #1.

### CODE REVIEW

**Comment by LuchoTurtle** [comment]
Kinda frustrating having to manually bump the `SDK` to version 21 because [`flutter_quill_extensions`](https://github.com/singerdmx/flutter-quill/tree/master/flutter_quill_extensions) is using a `gallery_saver` library that is tossing the error:

```
Launching lib/main.dart on ONEPLUS A6010 in debug mode...
main.dart:1
/Users/lucho/Documents/dwyl/flutter-wysiwyg-editor-tutorial/android/app/src/debug/AndroidManifest.xml Error:
	uses-sdk:minSdkVersion 19 cannot be smaller than version 21 declared in library [:gallery_saver] /Users/lucho/Documents/dwyl/flutter-wysiwyg-editor-tutorial/build/gallery_saver/intermediates/merged_manifest/debug/AndroidManifest.xml as the library might be using APIs not available in 19
	Suggestion: use a compatible library with a minSdk of at most 19,
		or increase this project's minSdk version to at least 21,
		or use tools:overrideLibrary="carnegietechnologies.gallery_saver" to force usage (may lead to runtime failures)
```

They should either upgrade or use another library, as stated in https://github.com/singerdmx/flutter-quill/issues/1350 :\

**Comment by nelsonic** [comment]
That _is_ frustrating. 🤦‍♂️ 
Are you able to `Fork` it and make the necessary updates then create a PR to update upstream repo?

**Comment by LuchoTurtle** [comment]
Perhaps but I think I can work with having `minSdkVersion` set to `21` (I just needed to upgrade from `20` to `21`, so it's not THAT big of a deal.

I'll see to it after sorting through this epic.

**Comment by LuchoTurtle** [comment]
Been trying to paste images directly into the clipboard and not from the gallery (as an optional route) but it doesn't work in Android devices (solely, it works on other platforms).
This is because of this line:

https://github.com/singerdmx/flutter-quill/blob/5dca6008cde707beb46291b6fbf71af6dbbf7bfb/lib/src/widgets/raw_editor.dart#L1534

Unfortunately, this is [not straightforward in Android devices](https://stackoverflow.com/questions/65432586/clipboard-an-image-in-flutter). I'll keep doing the "opening gallery and choosing" route.

**Comment by LuchoTurtle** [comment]
While this works on mobile, it does not work on `web`. I've explained one route that might get to work in https://github.com/singerdmx/flutter-quill/issues/976. 🙆‍♂️

**Comment by nelsonic** [comment]
@LuchoTurtle please consider opening a _separate_ issue for pasting images from clipboard for `Android` so that you can ship the basic version of this. I know you want to get _everything_ on the requirements working. But it's better to ship something _basic_ than get stuck on a "nice to have" feature like this. 👌 

Saw your comment on using `universal_io`. Seems like a no-brainer to use it. 💭

**Comment by LuchoTurtle** [comment]
Opened a PR that addresses an issue that I had embedding an image  -> https://github.com/singerdmx/flutter-quill/pull/1367.

Although it works, I'm testing this with an image URL, not a local file. I don't believe this is actually possible, so we'd need to upload an image to our `S3` bucket.

**Comment by LuchoTurtle** [comment]
I've dismissed the previous PR because it doesn't properly work on mobile devices.

Ugh, having this working on both `web` and `mobile` is utterly frustrating.

**Comment by nelsonic** [comment]
@LuchoTurtle please include a _lot_ more detail about your frustration. 🙏 
If something isn't working on a particular platform e.g. `Web` 🕸️ 
Please just open an issue and include screenshots, `GIFs` or `Video` explaining the issue. 🖼️ 
It might be that the insight you provide helps us to consider an alternative approach. 💭

**Comment by LuchoTurtle** [comment]
I've added more detail and following steps to the comment 👌

**Comment by LuchoTurtle** [comment]
I'm still in the process of writing the README detailing the steps needed to get everything working (even with the caveat that images on the web do not work as of now, they are **correctly embeded with a given image URL**, we just need to have it upload to a third-party service like our `imgup`).

Testing image uploads in `testWidgets`, as detailed in previous comments and in https://github.com/singerdmx/flutter-quill/issues/1389 have not been successful. 
I've tried c[hannel mocking](https://docs.flutter.dev/testing/plugins-in-tests#mock-the-platform-channel) (even though it's not advised, only as a last-minute resort). Even though Flutter advises to m[ock plugins](https://docs.flutter.dev/testing/testing-plugins) to test them, I can't do it because it's hidden under `flutter-quill`'s code.

I've only achieved 65% code coverage but this blocker doesn't allow me to go further.

Would it make sense to finish the README to have this working and give `visual-editor` a try in this same repo (although on a different PR)? Testing may be different and easier, as `visual-editor` is purportedly much better documented, contrary to `flutter-quill` which lacks tests.

**Comment by nelsonic** [comment]
@LuchoTurtle yeah, 65% is quite low. Perhaps it's worth having a small paragraph in the `README.md` explaining what _can't_ be tested with `flutter-quill` ... 

Images should _always_ be uploaded to the "3rd" party service and then included via `URL`.

**Comment by LuchoTurtle** [comment]
I've added the rest of the documentation and added all of the details into the `README` file.
After this, I've tried going through platform channels in a last attempt to override the behaviour from `image-picker` inside `flutter-quill` by using [`setMockMethodCallHandler`](https://stackoverflow.com/questions/52028969/testing-flutter-code-that-uses-a-plugin-and-platform-channel) and *although I got some progress by running the mocked overridden version*, I keep getting this following error:

```
The following PlatformException was thrown running a test:
PlatformException(error, Invalid argument: Instance of 'XFile', null, null)

When the exception was thrown, this was the stack:
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:652:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:310:18)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)
```

I feel like I've spent too much already on trying to get this to work, so I wager it makes sense to tackle uploading to `S3` and rendering on the web properly.

**Comment by LuchoTurtle** [comment]
Following https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/issues/1#issuecomment-1707962562, I'm submitting this for review.

It does not *yet* upload images to `S3`, as I believe this should be contained on a different PR and because this is already a big one to review.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,1655 @@
...
+this output on the terminal.
+
+```sh
+00:02 +1: All tests passed!   
+```
+
+This means everything is correctly set up!
+We are ready to start implementing!
+
+### **Make sure your `flutter` is up-to-date!
```
Comment (nelsonic) [inline README.md:123]:
```suggestion
### Make sure your `flutter` is up-to-date!
```

**Comment by nelsonic** [comment]
Just ran `flutter test --coverage` on this branch on my `localhost` got:

image

00:04 +0: loading /Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart                                                ../../.pub-cache/hosted/pub.dev/flutter_quill-7.4.1/lib/src/widgets/raw_editor.dart:353:7: Error: No named parameter with the name
'onLiveTextInput'.
      onLiveTextInput: null,
      ^^^^^^^^^^^^^^^
/opt/homebrew/Caskroom/flutter/3.3.8/flutter/packages/flutter/lib/src/widgets/editable_text.dart:1832:38: Context: Found this
candidate, but the arguments don't match.
  static List<ContextMenuButtonItem> getEditableButtonItems({
                                     ^^^^^^^^^^^^^^^^^^^^^^
00:08 +0 -1: loading /Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart [E]
  Failed to load "/Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart": Compilation failed for testPath=/Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart

To run this test again: /opt/homebrew/Caskroom/flutter/3.3.8/flutter/bin/cache/dart-sdk/bin/dart test /Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart -p vm --plain-name 'loading /Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart'
00:08 +0 -1: Some tests failed.

```sh
flutter doctor -v

[✓] Flutter (Channel stable, 3.10.6, on macOS 13.5.1 22G90 darwin-arm64, locale en-GB)
    • Flutter version 3.10.6 on channel stable at /opt/homebrew/Caskroom/flutter/3.3.8/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f468f3366c (9 weeks ago), 2023-07-12 15:19:05 -0700
    • Engine revision cdbeda788a
    • Dart version 3.0.6
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/n/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /Users/n/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.82.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.72.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.5.1 22G90 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 116.0.5845.187
    ! Error: Failed to prepare the device for development. Please unlock and reconnect the device. (code 806)

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

@LuchoTurtle please check. 🙏

**Comment by nelsonic** [comment]
Upgrading `Flutter` on my `localhost`:
```sh
flutter upgrade --force
Upgrading Flutter to 3.13.4 from 3.10.6 in /opt/homebrew/Caskroom/flutter/3.3.8/flutter...
```

To see if that's the problem. ⏳

**Comment by nelsonic** [comment]
```sh
Flutter 3.13.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 367f9ea16b (3 days ago) • 2023-09-12 23:27:53 -0500
Engine • revision 9064459a8b
Tools • Dart 3.1.2 • DevTools 2.25.0

Running flutter doctor...
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.13.4, on macOS 13.5.1 22G90 darwin-arm64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.3)
[✓] VS Code (version 1.82.0)
[✓] Connected device (2 available)
    ! Error: Failed to prepare the device for development. Please unlock and reconnect the device. (code 806)
[✓] Network resources

• No issues found!
```

Then:
```
flutter test
```

Output:

```sh
...
> vm_service 11.7.1 (was 11.3.0) (11.10.0 available)
+ web 0.1.4-beta
  win32 5.0.6 (5.0.7 available)
  xdg_directories 1.0.2 (1.0.3 available)
  xml 6.3.0 (6.4.2 available)
Changed 8 dependencies!
00:11 +3: All tests passed!
```

All good. ✅ 

image

🙄

**Comment by nelsonic** [comment]
Is it not _possible_ to test this file? 💭 
image

**Comment by nelsonic** [comment]
Converted your demo video to `GIF` and double-speed: https://ezgif.com/optimize/ezgif-2-26cec1fc55.gif

dwyl-wysiwyg-demo-optimised

**Comment by nelsonic** [comment]
Image uploads don't work on my `localhost` in `Chrome`:
image

This is "OK" as it will be fixed by using `visual-editor` in the follow-up. ✅

**Review by nelsonic — APPROVED** [review]
Thanks @LuchoTurtle :shipit:

**Pull request merged**